tests/MongoDB.Driver.Tests/Specifications/crud/CrudTestRunner.cs

```diff
@@ -19,55 +19,228 @@
 
             var arguments = (BsonDocument)operation.GetValue("arguments", new BsonDocument());
-            string reason;
-            if (!test.CanExecute(DriverTestConfiguration.Client.Cluster.Description, arguments, out reason))
+            test.SkipIfNotSupported(arguments);
+
+            test.Execute(client.Cluster.Description, database, collection, arguments, outcome, isErrorExpected, async);
+
+            if ((test.ActualError != null) != isErrorExpected)
             {
-                throw new SkipException(reason);
+                if (test.ActualError != null)
+                {
+                    throw test.ActualError;
+                }
+                else
+                {
+                    throw new Exception("The test was expected to throw an exception, but the executing was successful.");
+                }
             }
+        }
 
-            test.Execute(DriverTestConfiguration.Client.Cluster.Description, database, collection, arguments, outcome, async);
+        private IMongoCollection<BsonDocument> GetCollection(IMongoDatabase database, string collectionName, BsonDocument operation)
+        {
+            var collectionSettings = ParseCollectionOptions(operation);
+
+            var collection = database.GetCollection<BsonDocument>(
+                collectionName,
+                collectionSettings);
+
+            return collection;
+        }
+
+        private string GetCollectionName(BsonDocument operation)
+        {
+            if (operation.TryGetValue("collection_name", out var collectionName))
+            {
+                return collectionName.AsString;
+            }
+            else
+            {
+                return DriverTestConfiguration.CollectionNamespace.CollectionName;
+            }
+        }
+
+        private void ParseCollectionOptions(MongoCollectionSettings settings, BsonDocument collectionOptions)
+        {
+            foreach (var collectionOption in collectionOptions.Elements)
+            {
+                switch (collectionOption.Name)
+                {
+                    case "readConcern":
+                        settings.ReadConcern = ReadConcern.FromBsonDocument(collectionOption.Value.AsBsonDocument);
+                        break;
+                }
+            }
+        }
+
+        private MongoCollectionSettings ParseCollectionOptions(BsonDocument operation)
+        {
+            MongoCollectionSettings settings = null;
+            foreach (var operationElement in operation.Elements)
+            {
+                settings = settings ?? new MongoCollectionSettings();
+                switch (operationElement.Name)
+                {
+                    case "collectionOptions":
+                        ParseCollectionOptions(settings, operationElement.Value.AsBsonDocument);
+                        break;
+                }
+            }
+
+            return settings;
```

Comment:
I'm not sure why there is a foreach loop here. Wouldn't the following be simpler?

    if (operation.TryGetValue("collectionOptions", out var collectionOptions))
    {
        var settings = new MongoCollectionSettings();
        ParseCollectionOptions(settings, collectionOptions.AsBsonDocument);
        return settings;
    }
    else
    {
        return null;
    }

Reply:
Done